File: lib/pelican_installer/screens/menu.py
"""Main menu screen with component selection."""

from __future__ import annotations

from textual import on
from textual.app import ComposeResult
from textual.containers import Container
from textual.screen import Screen
from textual.widgets import Button, Static

from pelican_installer.components.menu import InstallerMenu
from pelican_installer.utils.state import InstallState


class MenuScreen(Screen[str]):
    """Main menu for selecting installation component."""

    CSS = """
    MenuScreen {
        align: center middle;
    }
    """

    def __init__(
        self,
        state: InstallState,
        panel_installed: bool = False,
        wings_installed: bool = False,
    ) -> None:
        super().__init__()
        self.state = state
        self.panel_installed = panel_installed
        self.wings_installed = wings_installed

    def compose(self) -> ComposeResult:
        with Container(id="root"):
            with Container(id="card"):
                yield Static("Pelican Panel Installer", id="title")
                yield Static(
                    f"OS: {self.state.os_name} {self.state.os_version}",
                    id="system-info",
                )
                yield InstallerMenu(id="main-menu")
                yield Static(
                    "Use ↑/↓, 1/2, Enter or click to select. Close: (c)",
                    id="hint",
                )
                with Container(id="footer"):
                    yield Button("Close (c)", id="close")

    def on_mount(self) -> None:
        """Set up menu options based on system state."""
        menu = self.query_one("#main-menu", InstallerMenu)

        # Build options based on what's installed
        options = []

        if not self.panel_installed:
            options.append("1) Install Panel")
        else:
            options.append("1) Update/Reinstall Panel")
            options.append("2) Uninstall Panel")

        if not self.wings_installed:
            idx = len(options) + 1
            options.append(f"{idx}) Install Wings")
        else:
            idx = len(options) + 1
            options.append(f"{idx}) Update/Reinstall Wings")
            options.append(f"{idx+1}) Uninstall Wings")

        menu.clear_options()
        for opt in options:
            menu.add_option(opt)

        if options:
            menu.highlighted = 0

    @on(InstallerMenu.OptionSelected)
    def handle_selection(self, event: InstallerMenu.OptionSelected) -> None:
        """Handle menu selection."""
        menu = self.query_one("#main-menu", InstallerMenu)
        selected = menu.get_option_at_index(event.option_index)
        if selected:
            text = str(selected.prompt).lower()

            # Parse selection
            if "install panel" in text and "uninstall" not in text:
                self.state.component = "panel"
                self.dismiss("webserver")
            elif "install wings" in text and "uninstall" not in text:
                self.state.component = "wings"
                self.dismiss("dependencies")
            elif "uninstall panel" in text:
                self.dismiss("uninstall_panel")
            elif "uninstall wings" in text:
                self.dismiss("uninstall_wings")
            elif "update" in text or "reinstall" in text:
                if "panel" in text:
                    self.state.component = "panel"
                    self.dismiss("webserver")
                elif "wings" in text:
                    self.state.component = "wings"
                    self.dismiss("dependencies")

    @on(Button.Pressed, "#close")
    def close_pressed(self) -> None:
        """Handle close button."""
        self.app.exit()

    def action_request_close(self) -> None:
        """Global close action (c key)."""
        self.app.exit()

File: src/screens/menu.rs
use crate::screens::{Screen, ScreenResult};
use crate::ui::components::SelectMenu;
use crate::utils::state::{InstallState, ComponentType};
use ratatui::prelude::*;
use ratatui::widgets::Paragraph;
use crossterm::event::KeyEvent;

pub struct MenuScreen {
    menu: SelectMenu,
}

impl MenuScreen {
    pub fn new(panel_installed: bool, wings_installed: bool) -> Self {
        let mut items = Vec::new();
        
        if !panel_installed {
            items.push("1) Install Panel".to_string());
        } else {
            items.push("1) Update/Reinstall Panel".to_string());
            items.push("2) Uninstall Panel".to_string());
        }
        
        if !wings_installed {
            items.push(format!("{}) Install Wings", items.len() + 1));
        } else {
            items.push(format!("{}) Update/Reinstall Wings", items.len() + 1));
            items.push(format!("{}) Uninstall Wings", items.len() + 1));
        }
        
        Self { menu: SelectMenu::new(items) }
    }
}

impl Screen for MenuScreen {
    fn render(&mut self, area: Rect, frame: &mut Frame, state: &InstallState) -> crate::error::Result<()> {
        let chunks = Layout::default()
            .direction(Direction::Vertical)
            .margin(1)
            .constraints([
                Constraint::Length(3),  // Title
                Constraint::Length(2),  // System info
                Constraint::Min(10),   // Menu
                Constraint::Length(2),  // Hint
                Constraint::Length(3),  // Footer
            ])
            .split(area);
        
        // Title
        let title = Paragraph::new("Pelican Panel Installer")
            .style(Style::default().fg(Color::Cyan).add_modifier(Modifier::BOLD))
            .alignment(Alignment::Center);
        frame.render_widget(title, chunks[0]);
        
        // System info
        let sys_info = format!("OS: {} {}", state.os_name, state.os_version);
        let info_para = Paragraph::new(sys_info)
            .style(Style::default().fg(Color::Gray))
            .alignment(Alignment::Center);
        frame.render_widget(info_para, chunks[1]);
        
        // Menu
        self.menu.render(chunks[2], frame.buffer_mut());
        
        // Hint
        let hint = Paragraph::new("Use ↑↓, 1-9, Enter to select | Close: c")
            .style(Style::default().fg(Color::Gray))
            .alignment(Alignment::Center);
        frame.render_widget(hint, chunks[3]);
        
        // Footer
        let footer = Paragraph::new("Press c to close")
            .style(Style::default().fg(Color::Yellow))
            .alignment(Alignment::Center);
        frame.render_widget(footer, chunks[4]);
        
        Ok(())
    }
    
    fn handle_key(&mut self, key: KeyEvent, state: &mut InstallState) -> crate::error::Result<ScreenResult> {
        use crossterm::event::{KeyCode, KeyModifiers};
        
        match key.code {
            KeyCode::Up | KeyCode::Char('k') => {
                self.menu.previous();
            }
            KeyCode::Down | KeyCode::Char('j') => {
                self.menu.next();
            }
            KeyCode::Enter => {
                let selected = self.menu.selected_text().ok_or_else(|| {
                    crate::error::InstallerError::InvalidInput("No item selected".to_string())
                })?;
                
                if selected.contains("install panel") && !selected.contains("uninstall") {
                    state.component = Some(ComponentType::Panel);
                    return Ok(ScreenResult::Next(Box::new(super::webserver::WebserverScreen::new())));
                } else if selected.contains("install wings") && !selected.contains("uninstall") {
                    state.component = Some(ComponentType::Wings);
                    return Ok(ScreenResult::Next(Box::new(super::install::InstallScreen::new())));
                } else if selected.contains("update") || selected.contains("reinstall") {
                    if selected.contains("panel") {
                        state.component = Some(ComponentType::Panel);
                        return Ok(ScreenResult::Next(Box::new(super::webserver::WebserverScreen::new())));
                    } else if selected.contains("wings") {
                        state.component = Some(ComponentType::Wings);
                        return Ok(ScreenResult::Next(Box::new(super::install::InstallScreen::new())));
                    }
                } else if selected.contains("uninstall") {
                    // TODO: Implement uninstall
                    return Ok(ScreenResult::Back);
                }
            }
            KeyCode::Char('c') => {
                return Ok(ScreenResult::Exit);
            }
            KeyCode::Char(c) if c.is_ascii_digit() => {
                let idx = c.to_digit(10).unwrap() as usize;
                if idx > 0 && idx <= self.menu.selected_index() + 1 {
                    self.menu.select_by_index(idx - 1);
                }
            }
            _ => {}
        }
        Ok(ScreenResult::None)
    }
}
